spec.bs

```diff
@@ -44,19 +58,81 @@ This specification extends the client-side storage available beyond cookies.
+```
 
-TBD
+Later on, Alex visits `https://video.example/`, which has an <{iframe}> on it which loads `https://social.example/heart-button`. In this case, the `social.example` {{Document}} |doc| is in a [=third party context=], and the local storage set previously might or might not be visible depending on User Agent storage access policies.
 
-<h4 id="cookies">Cookies</h4>
+Script in the <{iframe}> can call |doc|`.`{{Document/requestStorageAccess(types)}} to request access.
 
-TBD
+```javascript
+let handle = await document.requestStorageAccess({localStorage: true});
+let userid = handle.localStorage.getItem("userid");
+```
+
+</div>
+
+<h3 id="document-changes">Changes to {{Document}}</h3>
+
+<pre class="idl">
+dictionary StorageAccessTypes {
+  boolean all = false;
+  boolean cookies = false;
+};
+
+interface StorageAccessHandle {
+};
+
+partial interface Document {
+  Promise&lt;boolean> hasUnpartitionedCookieAccess();
+  Promise&lt;StorageAccessHandle> requestStorageAccess(StorageAccessTypes types);
+};
+</pre>
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasUnpartitionedCookieAccess()</code></dfn> method must run these steps:
+
+1. Invoke {{Document/hasStorageAccess()}}
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>requestStorageAccess(types)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |types|'s member |all| is `false` and |types|'s member |cookies| is `false`, then [=/reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. Let |requestUnpartitionedCookieAccess| be true if |types|'s member |all| or |types|'s member |cookies| is `true` and `false` otherwise.
+1. [=React=] |p| to the invocation of {{Document/requestStorageAccess()}} given |requestUnpartitionedCookieAccess| and invoke the following steps if the promise is fufilled (if the promise is rejected then apply that same rejection to |p|):
+    1. Let |handle| be a new object of type {{StorageAccessHandle}} with |types|.
+    1. [=/Resolve=] |p| with |handle|.
+1. Return |p|.
+
+<h3 id="has-storage-access-changes">Changes to {{Document/hasStorageAccess()}}</h3>
+
+Now that {{Document/requestStorageAccess(types)}} can be used to request [=unpartitioned data=] with or without specifically requesting cookies, it must be made clear that {{Document/hasStorageAccess()}} only returns true if [=first-party-site context=] cookies are accessable to the current document.
+As a function name, {{Document/hasUnpartitionedCookieAccess()}} more clearly communicates this.
+For now {{Document/hasStorageAccess()}} is not considered deprecated, but that may be worth taking up in future.
+
+<h3 id="request-storage-access-changes">Changes to {{Document/requestStorageAccess()}}</h3>
+
+Modify {{Document/requestStorageAccess()}} to take a `boolean` |requestUnpartitionedCookieAccess| as a non-web-exposed argument which defaults to true.
```

Comment:
I don't really understand this one, what's a non-web-exposed argument and why would we specify it here?